File: processor/src/main/java/io/michaelrocks/lightsaber/processor/descriptors/ProviderDescriptor.java
/*
 * Copyright 2015 Michael Rozumyanskiy
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.michaelrocks.lightsaber.processor.descriptors;

import io.michaelrocks.lightsaber.processor.annotations.AnnotationDescriptor;
import io.michaelrocks.lightsaber.processor.signature.TypeSignature;
import org.objectweb.asm.Type;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class ProviderDescriptor {
    private final Type providerType;
    private final QualifiedType providableType;
    private final FieldDescriptor providerField;
    private final MethodDescriptor providerMethod;
    private final Type moduleType;
    private final Type delegatorType;

    public ProviderDescriptor(final Type providerType, final QualifiedType providableType,
            final FieldDescriptor providerField, final Type moduleType) {
        this(providerType, providableType, providerField, null, moduleType, null);
    }

    public ProviderDescriptor(final Type providerType, final QualifiedType providableType,
            final MethodDescriptor providerMethod, final Type moduleType, final Type delegatorType) {
        this(providerType, providableType, null, providerMethod, moduleType, delegatorType);
    }

    private ProviderDescriptor(final Type providerType, final QualifiedType providableType,
            final FieldDescriptor providerField, final MethodDescriptor providerMethod, final Type moduleType,
            final Type delegatorType) {
        this.providerType = providerType;
        this.providableType = providableType;
        this.providerField = providerField;
        this.providerMethod = providerMethod;
        this.moduleType = moduleType;
        this.delegatorType = delegatorType;
    }

    public Type getProviderType() {
        return providerType;
    }

    public QualifiedType getQualifiedProvidableType() {
        return providableType;
    }

    public AnnotationDescriptor getQualifier() {
        return providableType.getQualifier();
    }

    public Type getProvidableType() {
        return providableType.getType();
    }

    public FieldDescriptor getProviderField() {
        return providerField;
    }

    public MethodDescriptor getProviderMethod() {
        return providerMethod;
    }

    public Type getModuleType() {
        return moduleType;
    }

    public Type getDelegatorType() {
        return delegatorType;
    }

    public List<QualifiedType> getDependencies() {
        if (providerMethod == null) {
            return Collections.emptyList();
        }

        final List<QualifiedType> dependencies = new ArrayList<>(providerMethod.getArgumentTypes().size());
        for (final TypeSignature argumentType : providerMethod.getArgumentTypes()) {
            final Type dependencyType = argumentType.getParameterType() != null
                    ? argumentType.getParameterType() : argumentType.getRawType();
            dependencies.add(new QualifiedType(dependencyType));
        }
        return dependencies;
    }
}

File: processor/src/main/kotlin/io/michaelrocks/lightsaber/processor/descriptors/ProviderDescriptor.kt
/*
 * Copyright 2016 Michael Rozumyanskiy
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package io.michaelrocks.lightsaber.processor.descriptors

import io.michaelrocks.grip.mirrors.AnnotationMirror
import io.michaelrocks.lightsaber.processor.commons.parameterType
import io.michaelrocks.lightsaber.processor.commons.rawType
import org.objectweb.asm.Type
import java.util.*

class ProviderDescriptor private constructor(
    val providerType: Type,
    val qualifiedProvidableType: QualifiedType,
    val providerField: QualifiedFieldDescriptor?,
    val providerMethod: QualifiedMethodDescriptor?,
    val moduleType: Type,
    val delegatorType: Type?
) {
  constructor(
      providerType: Type,
      providableType: QualifiedType,
      providerField: QualifiedFieldDescriptor,
      moduleType: Type
  ) : this(providerType, providableType, providerField, null, moduleType, null)

  constructor(
      providerType: Type,
      providableType: QualifiedType,
      providerMethod: QualifiedMethodDescriptor,
      moduleType: Type,
      delegatorType: Type?
  ) : this(providerType, providableType, null, providerMethod, moduleType, delegatorType)
}

val ProviderDescriptor.qualifier: AnnotationMirror?
  get() = qualifiedProvidableType.qualifier

val ProviderDescriptor.providableType: Type
  get() = qualifiedProvidableType.type

val ProviderDescriptor.dependencies: List<QualifiedType>
  get() {
    if (providerMethod == null) {
      return emptyList()
    }

    val dependencies = ArrayList<QualifiedType>(providerMethod.argumentTypes.size)
    for (index in 0..providerMethod.argumentTypes.size - 1) {
      val argumentType = providerMethod.argumentTypes[index]
      val dependencyType = argumentType.parameterType ?: argumentType.rawType
      val qualifier = providerMethod.parameterQualifiers[index]
      dependencies.add(QualifiedType(dependencyType, qualifier))
    }
    return dependencies
  }

val ProviderDescriptor.isConstructorProvider: Boolean
  get() = providerMethod != null && providerMethod.isConstructor
